[XML Parser] Don't overwrite result of xmlGetPredefinedEntity

zakariaridouh opened the pull request:
#### f337e8735a7df410fa90a1a34d0f9a18f39ed39b
<pre>
[XML Parser] Don&apos;t overwrite result of xmlGetPredefinedEntity
<a href="https://bugs.webkit.org/show_bug.cgi?id=288153">https://bugs.webkit.org/show_bug.cgi?id=288153</a>
<a href="https://rdar.apple.com/145249951">rdar://145249951</a>

Reviewed by David Kilzer.

Currently, in `xmlGetPredefinedEntity` we overwrite the value of `ent-&gt;etype`,
which should always be of type `XML_INTERNAL_PREDEFINED_ENTITY`

We should use an ASSERT to verify that the result has the expected
type, and not overwrite it.

This allows libxml2 to change these variables to `const` without us
being impacted.

* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::getEntityHandler):

Canonical link: <a href="https://commits.webkit.org/290765@main">https://commits.webkit.org/290765@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25b6829a95621d9c2af10a78d16ee17c1777c522

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91006 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96034 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10947 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18867 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69968 "15 flakes 95 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27487 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94007 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82462 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50294 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8098 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40925 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/78399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18208 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18468 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78291 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78172 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19315 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22656 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18215 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17950 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21409 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19735 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
```diff
@@ -1259,7 +1259,7 @@ static xmlEntityPtr getEntityHandler(void* closure, const xmlChar* name)
     xmlEntityPtr ent = xmlGetPredefinedEntity(name);
     if (ent) {
-        ent->etype = XML_INTERNAL_PREDEFINED_ENTITY;
+        ASSERT(ent->etype == XML_INTERNAL_PREDEFINED_ENTITY);
```
Please use `RELEASE_ASSERT` here to catch real bugs outside of Debug EWS bots.

zakariaridouh replied:
Fixed.

ddkilzer reviewed (approved):
r=me if you switch to `RELEASE_ASSERT()`.

zakariaridouh committed:
[XML Parser] Don't overwrite result of xmlGetPredefinedEntity
https://bugs.webkit.org/show_bug.cgi?id=288153
rdar://145249951

Reviewed by David Kilzer.

Currently, in `xmlGetPredefinedEntity` we overwrite the value of `ent->etype`,
which should always be of type `XML_INTERNAL_PREDEFINED_ENTITY`

We should use an ASSERT to verify that the result has the expected
type, and not overwrite it.

This allows libxml2 to change these variables to `const` without us
being impacted.

* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::getEntityHandler):

Canonical link: https://commits.webkit.org/290765@main
